[Core] CLI flags, initial selfbot support, revamped setup / main screen, etc

Twentysix26 opened the pull request:
**Changes:**
Startup flags: can be started and configured without ever doing the interactive setup
Changed default settings format so that all unset values are None
Removed new cogs prompt
Removed `LOGIN_TYPE` from settings.json. It now defaults to token and fallbacks to email/password
Smarter initial setup: only asks for the settings that are actually missing
Startup flag that allows settings to be memory-only
Initial selfbot support
Only reset login credentials (on confirmation) instead of deleting the whole file in case of login failure
Revamped main screen
Made sure that nothing blows up when you run Red on Windows without `chcp 65001`
Possibility of setting credentials in the environment variables `RED_TOKEN` / `RED_EMAIL` `RED_PASSWORD` 

**Possible breaking changes:**
Cogs that make use of bot.email and expect a token (I don't think there are any, but still)
Cogs that set settings and expect them to save automatically. This is now disabled to allow memory-only settings.
`Settings.login_type` is no longer a thing

**Warning:** Testing this PR will change the format of your settings.json. It won't work anymore if you try to use it again in the current build.

Twentysix26 committed:
[Core] Startup flags, vanilla cogs enabled by default, revamped setup

Changed default settings format so that all unset values are None
Removed new cogs prompt
Smarter initial setup: only asks for the settings that are actually missing
Startup flag that allows settings to be memory-only
Initial selfbot support
Only reset login credentials (on confirmation) instead of deleting the whole file in case of login failure

tekulvw commented:
Can we get a debug cli flag?

Twentysix26 commented:
@tekulvw desired behaviour? Othen than a `Settings.debug` being set to True, which would be useful for cogs as a debug mode flag

Kowlin commented:
@Twentysix26 pretty sure we talked about this before. the Debug flag was to enable the debug command, with the debug command defaulting to disabled.

tekulvw commented:
Uh...well...for starters I'd set the `red` logger to `logging.DEBUG` then keep the flag available in settings for cog loggers to do the same (most of the time they shouldn't as they just inherit the logging level from `red`).

@Kowlin not sure where that came from tbh

edit: Use `red.mod` as the logger name in Mod, also `red.discordpy` for the dpy logger?

Twentysix26 committed:
Added debug mode flag

Twentysix26 commented:
@tekulvw I assume d.py's logger has to be named `discord`
https://discordpy.readthedocs.io/en/latest/logging.html

Twentysix26 committed:
Edits to allow selfbots to function

Twentysix26 committed:
Revamped main screen, show owner, set selfbot's owner id

calebj commented on red.py:
```diff
@@ -129,7 +129,10 @@ def clear_message_modifiers(self):
+        if author.bot:
+            return False
+
+        if author == self.user and not self.settings.self_bot:
```
Change this to
```py
if author == self.user:
    return self.settings.self_bot
```

Twentysix26 committed:
Fixed check

tekulvw commented on cogs/mod.py:
```diff
@@ -1378,7 +1378,7 @@ def setup(bot):
     check_folders()
     check_files()
-    logger = logging.getLogger("mod")
+    logger = logging.getLogger("red.mod")
```
👍

tekulvw commented on cogs/utils/settings.py:
```diff
@@ -11,27 +14,96 @@ def __init__(self, path=default_path):
+
+    def parse_cmd_arguments(self):
+        parser = argparse.ArgumentParser(description="Red - Discord Bot")
+        parser.add_argument("--email")
```
Can you add help doc for this? Don't want people using user accounts

tekulvw commented on cogs/utils/settings.py:
```diff
@@ -11,27 +14,96 @@ def __init__(self, path=default_path):
+                self.prefixes = sorted(args.prefix, reverse=True)
+            elif arg == "admin_role":
+                self.default_admin = args.admin_role
+            elif arg == "mod_role":
```
Uh...why aren't you just using `args.var_name`? and why the `elif`'s?

Twentysix26 replied:
I swear it made sense when I was looking at how argparse worked. Refactoring :')

tekulvw reviewed (commented):
Haven't tested it, just first glance comments

Twentysix26 committed:
Added CLI flags help strings, refactored parsing

Twentysix26 committed:
Deal with UnicodeEncodeError when printing to console

Kowlin reviewed (commented):
Changes work fine, nothing broke, no buildings collapsed. And no demons are summoned (yet). One traceback when using ``--self-bot``

```
Ignoring exception in on_message
Traceback (most recent call last):
  File "C:\Users\K\AppData\Local\Programs\Python\Python35\lib\site-packages\discord\client.py", line 307, in _run_event
    yield from getattr(self, event)(*args, **kwargs)
  File "red.py", line 261, in on_message
    if user_allowed(message):
  File "red.py", line 152, in user_allowed
    for r in results:
  File "red.py", line 150, in <lambda>
    lambda name: discord.utils.get(author.roles, name=name),
AttributeError: 'User' object has no attribute 'roles'
```

Twentysix26 committed:
Allow login credentials in environment variables

Passing credentials through CLI flags is insecure.
If credentials are set in `RED_TOKEN` / `RED_EMAIL` `RED_PASSWORD` they will override settings.

Twentysix26 committed:
A sane(r) way to deal with settings

Removed LOGIN_TYPE, it was unnecessary. If a token is set it defaults to it, otherwise email/password are used.
Environment variables override saved settings and are now not saved to disk, as it should be.

Twentysix26 committed:
Correct login time, fixed issue with logger

Twentysix26 committed:
Set up the loggers first
